### PR TITLE
Added new steps

### PR DESCRIPTION
**Pull request by aalexandrasilva**
- Restructured the project. Mostly organized the files to the packages by component. For example, all claim checkers should be inside the package `claim_worthiness_check`.
- Added coreference resolution as a REST service with spring.
- Integrated coreference resolution with our pipeline. It is now between the translation and the claim check step.
- Added multiple configurations that we attempted to the first step of WISE such as scaling and over-sampling from minority classes.
- Added the RNN training script and integrated the trained model for the final prediction.
- Changed the output format to a better-formatted JSON.
- Added the result of the indicator check to our output.
- We also now create the database if it isn't there when starting the service, instead of running 2 separate scripts.

In [this file](https://github.com/dice-group/NEBULA/files/13615180/example_nela.txt), there's an example output of the service for `/rawstatus` and [this file](https://github.com/dice-group/NEBULA/files/13615186/example_nela_simple.txt)

 there is the output for `/status`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,7 +28,7 @@ http://127.0.0.1:5000
 
 __Methods__
 
-The Nebula application provides three methods: **check** ,**status**, **test or default**.
+The Nebula application provides three methods: **check** ,**status**,**rawstatus**, **test or default**.
```
Comment (MichaelRoeder) [inline README.md]: three... okay, maybe four :laughing:

File: README.md [hunk 2]
```diff
@@ -37,7 +37,7 @@ For POST requests, the text [text] input and optional language input [lang] shou
 Below are some examples of the POST request.
 
 ```shell
-curl http://nebulavm.cs.upb.de/check --header 'Content-Type: application/json' --data '{"lang": "en", "text": "Text I want to check"}'
+curl http://nebulavm.cs.upb.de/check --header 'Content-Type: application/json' --data '{"lang": "en", "text": "Text I want to check", "token": "device_token", "translation": "translated_text"}'
```
Comment (MichaelRoeder) [inline README.md]: Shouldn't we document that the device token as well as the translation are not mandatory? I actually would suggest to remove them from this line and add a short description of additional variables.

File: README.md [hunk 3]
```diff
@@ -46,7 +46,9 @@ import requests
 CHECK_URL = "http://nebulavm.cs.upb.de/check"
 input = {
     'lang': 'en',
-    'text': 'Text we want to check'
+    'text': 'Text we want to check',
+    'token': 'device_token',
+    'translation': 'translated_text'
```
Comment (MichaelRoeder) [inline README.md]: Same as above. The examples typically show the simplest way to call the service, which would be without a device token and translation.

File: README.md [hunk 4]
```diff
@@ -77,7 +79,16 @@ The check method will return a JSON object containing an ID for the text input.
 }
 ```
 
-**Status Method**
+Input parameters | Description | 
+--- | --- | 
+_lang_ | Language of the input text. If the language is not set to 'en' and not translation is provided, NEBULA will first translate it to english. The default is 'nd'.  |
+_text_ | The text we want to verify the veracity for. |
+_token_ | The device token to notify of an available fact-checking result. |
+_translation_ | The english translation of the input text. Used if we want to skip the translation step. |
```
Comment (MichaelRoeder) [inline README.md]: Okay, here are the parameter descriptions that I was looking for. It would be good to clearly mark which of them are mandatory, which are optional and which have default values (e.g., this is done for the language but can be done much shorter). You can have a look at a nearly random example API like https://maven.apache.org/plugins/maven-javadoc-plugin/javadoc-mojo.html
It doesn't have to be exactly as they do it. We do not really need the versioning part and maybe also not the data type column. However, I like that they found a structure that clearly telly you a) what is mandatory and b) what are default values.

File: README.md [hunk 5]
```diff
@@ -86,7 +97,7 @@ The status method will return a JSON object containing information on the status
 
 ```
 {
-    "id": "http://nebula.cs.uni-paderborn.de/id/123",
+    "id": "123",
```
Comment (MichaelRoeder) [inline README.md]: Oh no... my semantic web heart cannot handle the replacement of a URL with a numeric idea :broken_heart: 

But I guess it is okay in this setup :thinking:

File: org/diceresearch/nebula/data/dataset.py [hunk 6]
```diff
@@ -29,72 +26,154 @@ class StanceDataset(Dataset):
...
             self.label = kwargs.get('label')
         else:
             self.claim_id = list()
-            self.evidence_ids = list()  # TODO we probably need to add this to keep track
             self.stance_scores = list()
             self.label = list()
 
-            p_bar = tqdm(jsonl)
-            for item in p_bar:
+            resample = kwargs.get('resample')
```
Comment (MichaelRoeder) [inline org/diceresearch/nebula/data/dataset.py:45]: You documented the arguments so nicely further above. But "resample" is not mentioned there.

File: org/diceresearch/nebula/main.py [hunk 7]
```diff
@@ -54,63 +60,45 @@ def check():
...
 
+    # retrieve arguments
     text = args.get('text')
+    translated_text = args.get('translation')
     lang = args.get('lang')
     registration_token = args.get('token')
 
     # Assign not defined if language is not specified
     if not lang:
         lang = "nd"
```
Comment (MichaelRoeder) [inline org/diceresearch/nebula/main.py]: The readme stated that we will assume English... this is slightly confusing :thinking:

File: org/diceresearch/nebula/orchestrator.py [hunk 8]
```diff
@@ -2,140 +2,141 @@
...
-    INPUT_LANG = current[3]
-    TRANSLATED_TEXT = current[4]
-    CLAIM_CHECK_WORTHINESS_RESULT = current[6]
-    EVIDENCE_RETRIEVAL_RESULT = current[8]
-    STANCE_DETECTION_RESULT = current[10]
-
-    #notification
-    REGISTRATION_TOKEN = current[14]
+
+    stage_number = current[1]
```
Comment (MichaelRoeder) [inline org/diceresearch/nebula/orchestrator.py]: Using these IDs looks a little bit error prone to me... I am not sure whether we want to stick to it or whether we should look for an alternative :thinking: 

Maybe that is something that we want to discuss :slightly_smiling_face:

File: org/diceresearch/nebula/orchestrator.py [hunk 9]
```diff
@@ -2,140 +2,141 @@
...
-
-    elif next_stage == 7:
-        pass
-    elif next_stage == 8:
-        pass
-    elif next_stage == 9:
+    elif next_stage == 10:
         pass
     else:
         pass
```
Comment (MichaelRoeder) [inline org/diceresearch/nebula/orchestrator.py]: `elif next_stage == 10:` shouldn't exist since it is not used at all.

The `else:` part should throw an exception or at least log an error. If we end up in this block, something went wrong with our implementation. Typically, we want to know that :wink:

File: org/diceresearch/nebula/orchestrator.py [hunk 10]
```diff
@@ -2,140 +2,141 @@
...
+        # set status to ongoing and set timestamp
         databasemanager.update_step(settings.results_table_name, settings.status, settings.ongoing, identifier)
         databasemanager.update_step(settings.results_table_name, settings.timestamp, datetime.now().isoformat('#'),
                                     identifier)
-        # translate
-        # language
-        if INPUT_LANG != "en":
+
+        # translate if language differs from english and we were not provided a translation already
+        if input_lang != "en" and not translated_text:
```
Comment (MichaelRoeder) [inline org/diceresearch/nebula/orchestrator.py:64]: I think this collides with the previously defined default value "nd". The translation step may try to translate from not defined to English, or? :thinking:

File: org/diceresearch/nebula/orchestrator.py [hunk 11]
```diff
@@ -2,140 +2,141 @@
...
+                                        indicator_json, identifier)
+        else:
+            # otherwise show skipped status
+            status = settings.skipped
+
+        # update and increase the step
+        databasemanager.update_step(settings.results_table_name, settings.results_indicator_check_status,
+                                    status, identifier)
+        databasemanager.increase_the_stage(settings.results_table_name, identifier)
+        goNextLevel(identifier)
```
Comment (MichaelRoeder) [inline org/diceresearch/nebula/orchestrator.py:123]: We don't know what our partners want to do in this indicator check in the future. Since we already have the thread-based approach for the other steps, I would suggest to use it here as well. At least to me, that seems to be the saver way, just in case the indicator check takes some more time in the future. :thinking: 

On the other hand: I am wondering whether we even need all these threads that we start... Wouldn't it be sufficient to do it once at the very beginning for a newly created job? To me, it looks like one thread 1 is created, does its job, creates thread 2, terminates and than thread 2 runs and performs the next step. This is a little bit costly, when thread 1 could perform the steps sequentially :thinking: 

Maybe we should discuss this as :slightly_smiling_face:

File: org/diceresearch/nebula/tools/cross_validation.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+    else:
+        seed = random.randint(0, 1e6)
+
+    oversampler = RandomOverSampler(sampling_strategy='auto', random_state=seed)
+    train_dataset = StanceDataset(jsonl=training_data, k=10, resample=oversampler)
+
+    logging.debug('Seed: {}'.format(seed))
+    torch.manual_seed(seed)
+
+    # Grid search
```
Comment (MichaelRoeder) [inline org/diceresearch/nebula/tools/cross_validation.py:55]: Maybe a Bayesian Optimizer would be better than a grid search? :thinking:

File: org/diceresearch/nebula/tools/run_train_wise.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    # Print the best thresholds and F1 score
+    predicted_labels = [translate_to_classes(score, best_thresholds[0], best_thresholds[1], class_labels)
+                        for score in predicted_scores]
+
+    # calculate metrics for best thresholds
+    calculate_metrics(true_labels, predicted_labels, class_labels)
+
+
+def translate(label):
+    # FIXME
```
Comment (MichaelRoeder) [inline org/diceresearch/nebula/tools/run_train_wise.py]: What exactly needs to be fixed here? :laughing:

File: pom.xml [hunk 14]
```diff
@@ -0,0 +1,122 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0"
```
Comment (MichaelRoeder) [inline pom.xml:1]: We do not really need this Java service, or? We can simply use a dedicated server of the CoreNLP project (https://stanfordnlp.github.io/CoreNLP/corenlp-server.html#dedicated-server).

**Review by MichaelRoeder — COMMENTED** [review]
A lot of nice changes. However, I couldn't resist to add some comments :wink:

File: org/diceresearch/nebula/data/dataset.py (hunk 6 above)
Comment (aalexandrasilva) [reply org/diceresearch/nebula/data/dataset.py:45]: This argument is not connected to the service. For the service, resample would be `None`. At least not directly, resampling parameter is used while training the MLP, which takes place by running the run_train.py script: https://github.com/dice-group/NEBULA/blob/1488341168b5fd92fcb50d6f63fa6112dca55d6e/org/diceresearch/nebula/tools/run_train.py#L55-L64

Comment (aalexandrasilva) [reply org/diceresearch/nebula/data/dataset.py:45]: But maybe I should add to the README how to run the training scripts as well, and the parameters for these 🤔

File: org/diceresearch/nebula/main.py (hunk 7 above)
Comment (aalexandrasilva) [reply org/diceresearch/nebula/main.py]: The README states the default is 'nd' 🤔 If the language is not provided, we try to translate it to english.

File: org/diceresearch/nebula/orchestrator.py (hunk 8 above)
Comment (aalexandrasilva) [reply org/diceresearch/nebula/orchestrator.py:28]: That is true, maybe I can change the SELECT query to return a json object instead? It would save us from having to change these indices every time we modify the database.

File: org/diceresearch/nebula/orchestrator.py (hunk 10 above)
Comment (aalexandrasilva) [reply org/diceresearch/nebula/orchestrator.py:64]: How does it collide? 

> The translation step may try to translate from not defined to English, or?

Shouldn't that be the case? We need to have the text in english for the next steps.

File: pom.xml (hunk 14 above)
Comment (aalexandrasilva) [reply pom.xml:1]: As far as I remember from what @umairq mentioned, we needed it because CoreNLP only finds the coreference chains and doesn't replace them. This is the coreference resolution that was present in the FactCheckUtilityTools repository.
Maybe we can also use their dedicated server and do the replacement in the python side. This is only worth it if we insist on keeping Stanford's coreference. We also have the spacy implementation now in https://github.com/dice-group/NEBULA/blob/wise/org/diceresearch/nebula/coref_resolution/spacy_coref.py.
But maybe @umairq can shed some light there.

File: org/diceresearch/nebula/orchestrator.py (hunk 11 above)
Comment (aalexandrasilva) [reply org/diceresearch/nebula/orchestrator.py:123]: You are right, I've removed all the extra threads and it seems to work fine so far 🙂

File: org/diceresearch/nebula/orchestrator.py (hunk 8 above)
Comment (aalexandrasilva) [reply org/diceresearch/nebula/orchestrator.py]: I've modified it to return a dictionary. So we can access them by column name now: 
`stage_number = current["STAGE_NUMBER"]`
